activerecord/lib/active_record/relation.rb

```diff
@@ -304,84 +304,151 @@ def many?
       query_signature = ActiveSupport::Digest.hexdigest(to_sql)
       key = "#{klass.model_name.cache_key}/query-#{query_signature}"
 
-      if cache_version(timestamp_column)
+      if cache_version(id_column: id_column, timestamp_column: timestamp_column)
         key
       else
-        "#{key}-#{compute_cache_version(timestamp_column)}"
+        "#{key}-#{compute_cache_version(id_column: id_column, timestamp_column: timestamp_column)}"
       end
     end
     private :compute_cache_key
 
     # Returns a cache version that can be used together with the cache key to form
-    # a recyclable caching scheme. The cache version is built with the number of records
-    # matching the query, and the timestamp of the last updated record. When a new record
-    # comes to match the query, or any of the existing records is updated or deleted,
-    # the cache version changes.
+    # a recyclable caching scheme. The cache version is built with the number of records|
+    # matching the query, the IDs of the query's first and last records (if the query
+    # has a limit or offset and you're using MySQL or a version of PostgreSQL or SQLite that
+    # supports window functions), and the timestamp of the last updated record. When a
+    # new record comes to match the query, or any of the existing records is updated or
+    # deleted, the cache version changes.
     #
     # If the collection is loaded, the method will iterate through the records
     # to generate the timestamp, otherwise it will trigger one SQL query like:
     #
     #    SELECT COUNT(*), MAX("products"."updated_at") FROM "products" WHERE (name like '%Cosmic Encounter%')
-    def cache_version(timestamp_column = :updated_at)
+    def cache_version(deprecated_timestamp_column = nil, id_column: primary_key, timestamp_column: :updated_at)
+      if deprecated_timestamp_column
+        ActiveSupport::Deprecation.warn("Passing a timestamp column as a positional argument to ActiveRecord::Relation#cache_version is deprecated. Please use a `:timestamp_column` keyword argument instead.")
+        timestamp_column = deprecated_timestamp_column
+      end
+
       if collection_cache_versioning
         @cache_versions ||= {}
-        @cache_versions[timestamp_column] ||= compute_cache_version(timestamp_column)
+        @cache_versions[[id_column, timestamp_column]] ||= compute_cache_version(id_column: id_column, timestamp_column: timestamp_column)
       end
     end
 
-    def compute_cache_version(timestamp_column) # :nodoc:
+    def compute_cache_version(id_column: primary_key, timestamp_column: :updated_at) # :nodoc:
+      size = 0
+      first_id = nil
+      last_id = nil
+      timestamp = nil
+
       if loaded? || distinct_value
         size = records.size
+
         if size > 0
-          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column)
+          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column).utc.to_s(cache_timestamp_format)
```

Comment:
I see that the previous code used the timestamp as-is.  Does converting the timestamp to UTC fix another issue?

Reply:
The previous code did convert to UTC, just on a different line ([line 381 of relation.rb](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation.rb#L381)).

Not certain the reasoning behind the UTC conversion. I looked at the [original pull request implementing cache keys](https://github.com/rails/rails/pull/20884), and couldn't find an explanation. Perhaps it avoids the possibility mismatching keys if one's Rails app & database use different time zones.